### PR TITLE
Annotate Notification related endpoints with feature flags

### PR DESCRIPTION
**Pull request by darryl-apple**
#### e1bc22c3d8c4ffeefdaa7d57ec50ad2358d8b977
<pre>
Annotate Notification related endpoints with feature flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=283731">https://bugs.webkit.org/show_bug.cgi?id=283731</a>
<a href="https://rdar.apple.com/140592973">rdar://140592973</a>

Reviewed by Sihui Liu.

Annotate Notification related endpoints with feature flags

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::connectionToWebProcess):
(WebKit::NetworkNotificationManager::sharedPreferencesForWebProcess const):
(WebKit::NetworkNotificationManager::clearNotifications): Deleted.
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::sharedPreferencesForWebProcess const):
(WebKit::ServiceWorkerNotificationHandler::clearNotifications): Deleted.
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::sharedPreferencesForWebProcess const):
(WebKit::WebNotificationManagerMessageHandler::clearNotifications): Deleted.
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPreferences.mm:
(DisableRichJavaScriptFeatures)):

Canonical link: <a href="https://commits.webkit.org/287881@main">https://commits.webkit.org/287881@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/fd7c554242beadbe2646a843cc72435edaa52232

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84283 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30766 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81860 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7041 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62341 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20182 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82820 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42645 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26779 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29206 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72785 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70868 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27245 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85708 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78875 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4887 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70600 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7161 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68469 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70152 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17659 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13841 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12756 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101220 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6944 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12549 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24763 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6812 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10316 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8616 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in [hunk 1]
```diff
@@ -20,14 +20,18 @@
...
-    ShowNotification(struct WebCore::NotificationData notificationData, RefPtr<WebCore::NotificationResources> notificationResources) -> ()
-    CancelNotification(WebCore::SecurityOriginData origin, WTF::UUID notificationID)
-    ClearNotifications(Vector<WTF::UUID> notificationIDs)
-    DidDestroyNotification(WTF::UUID notificationID)
-    PageWasNotifiedOfNotificationPermission()
-    RequestPermission(WebCore::SecurityOriginData origin) -> (bool granted)
-    SetAppBadge(WebCore::SecurityOriginData origin, std::optional<uint64_t> badge)
-    GetPermissionState(WebCore::SecurityOriginData origin) -> (enum:uint8_t WebCore::PushPermissionState permission)
-    GetPermissionStateSync(WebCore::SecurityOriginData origin) -> (enum:uint8_t WebCore::PushPermissionState permission) Synchronous
+#if ENABLE(NOTIFICATIONS)
```
Comment (szewai) [inline Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in]: Why do we add the compile time flag here? Generally it should match declarations in `NotificationManagerMessageHandler`, but I don't see the flag there.

File: Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in [hunk 2]
```diff
@@ -20,14 +20,18 @@
...
+    [EnabledBy=NotificationsEnabled] ShowNotification(struct WebCore::NotificationData notificationData, RefPtr<WebCore::NotificationResources> notificationResources) -> ()
+    [EnabledBy=NotificationsEnabled] CancelNotification(WebCore::SecurityOriginData origin, WTF::UUID notificationID)
+    [EnabledBy=NotificationsEnabled] DidDestroyNotification(WTF::UUID notificationID)
+    [EnabledBy=NotificationsEnabled] PageWasNotifiedOfNotificationPermission()
+    [EnabledBy=NotificationsEnabled] RequestPermission(WebCore::SecurityOriginData origin) -> (bool granted)
+#endif
+#if ENABLE(WEB_PUSH_NOTIFICATIONS)
```
Comment (szewai) [inline Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:31]: Ditto.

File: Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h [hunk 3]
```diff
@@ -45,13 +46,13 @@ class NotificationManagerMessageHandler : public IPC::MessageReceiver {
 
     virtual void showNotification(IPC::Connection&, const WebCore::NotificationData&, RefPtr<WebCore::NotificationResources>&&, CompletionHandler<void()>&&) = 0;
     virtual void cancelNotification(WebCore::SecurityOriginData&&, const WTF::UUID& notificationID) = 0;
-    virtual void clearNotifications(const Vector<WTF::UUID>& notificationIDs) = 0;
```
Comment (szewai) [inline Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:48]: Why do we remove this? Is this because it's unused? If so, can we do the removal in a separate PR?

File: Source/WebKit/NetworkProcess/NetworkProcess.cpp [hunk 4]
```diff
@@ -2882,6 +2882,15 @@ NetworkConnectionToWebProcess* NetworkProcess::webProcessConnection(ProcessIdent
     return m_webProcessConnections.get(identifier);
 }
 
+NetworkConnectionToWebProcess* NetworkProcess::webProcessConnection(const IPC::Connection& connection) const
+{
+    for (Ref<NetworkConnectionToWebProcess> con : m_webProcessConnections.values()) {
```
Comment (szewai) [inline Source/WebKit/NetworkProcess/NetworkProcess.cpp]: Let's use `Ref webProcessConnection` for clarity (and you could omit the type)

**Review by szewai — APPROVED** [review]
r=me if bots are happy

File: Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in (hunk 2 above)
Comment (darryl-apple) [reply Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:31]: I've removed the #if ENABLE(NOTIFICATIONS) from the messages above, however the #if ENABLE(WEB_PUSH_NOTIFICATIONS) is needed as the definition of the BuiltInNotificationsEnabled flag is wrapped by it

File: Source/WebKit/NetworkProcess/NetworkProcess.cpp (hunk 4 above)
Comment (darryl-apple) [reply Source/WebKit/NetworkProcess/NetworkProcess.cpp]: Done

File: Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in (hunk 1 above)
Comment (rkirsling) [reply Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in]: Not having this broke the `!ENABLE(NOTIFICATIONS)` build, as the `notificationsEnabled` preference doesn't exist without `ENABLE(NOTIFICATIONS)`. This is clearly visible in UnifiedWebPreferences.yaml. Fixing in #38074.
